Extend semantic conventions for RPC and allow non-gRPC calls (#604)
open-telemetry/opentelemetry-specification: specification/trace/semantic_conventions/rpc.md

<!-- toc -->

- [Common remote procedure call conventions](#common-remote-procedure-call-conventions)
  * [Span name](#span-name)
  * [Attributes](#attributes)
- [gRPC](#grpc)
  * [Status](#status)
  * [Events](#events)

<!-- tocstop -->

## Common remote procedure call conventions

A remote procedure calls is described by two separate spans, one on the client-side and one on the server-side.
For outgoing requests, the `SpanKind` MUST be set to `CLIENT` and for incoming requests to `SERVER`.

### Span name

> The _span name_ MUST be the full RPC method name formatted as:

Reviewer: ```suggestion
The _span name_ MUST be the full RPC method name, when available, formatted as:
```

An HTTP REST call is also a form of RPC, but there may not be a formal "RPC method name"

Author: Can you think of some guidance of what to use if there is no proper name available?

Reviewer: @Oberon00 sgtm, the doc should spell out then what the scope of "RPC use case" is.

```

> $package.$service/$method

Reviewer 2: I guess some service won't have package so clarify it's optional?

Reviewer 3: I feel like the span name is still too Go+gRPC heavy.

Reviewer: internally at Uber our convention is `Service::Method`, without package, used to Thrift and Proto based RPCs. However, I can imaging the situations where more than one RPC interface is hosted in the same process whose only distinction is in the package name, e.g. `api.v1.Service::Method` and `api.v2.Service::Method`. So sgtm to include the package.

@Oberon00 why Go?  

Given the comment above to ignore HTTP REST calls, typical RPC frameworks use some form of IDL, where package/interface/method are pretty standard. I don't see it as gRPC-heavy.

Reviewer 2: Thanks - LGTM

Reviewer 3: The slashes made me think of go, but I guess you are right that this can be used for anything.

Reviewer: `::` instead of slash might result in a more intuitive string. Both are bad for passing span name in the URL query (require escaping).

```

(where $service must not contain dots).

Examples of span names:

- `grpc.test.EchoService/Echo`
- `com.example.ExampleRmiService/exampleMethod`

> - `MyCalcService.Calculator/Add` reported by the server and  

Reviewer 2: This example is confusing since the word `Service` is in the `package`. To my other point of making it optional, maybe we can drop the concept of package? Is there any intent of treating this token after the last dot specially?

Author: I think it's not unlikely that the term `service` appears somewhere in the package if someone implemented the app or library with the fact in mind that it will be exposed as a service.

Making it optional sounds reasonable to me but I would not drop it, the qualifier could still be valuable information.

Author: I see. In the PR description I actually asked if we should just include the package name in the `service.name` attribute, where this is currently explicitly excluded (and I didn't touch that definition). It would be fine with me to just pass the fully-qualified name here and not making the distinction (which is hard to do right for all languages anyway).

  `MyServiceReference.ICalculator/Add` reported by the client for .NET WCF calls

### Attributes

| Attribute name |                          Notes and examples                            | Required? |
| -------------- | ---------------------------------------------------------------------- | --------- |
| `rpc.system`   | A string identifying the remoting system, e.g., `"grpc"`, `"java_rmi"` or `"wcf"`.       | Yes |

> | `rpc.service`  | The service name, must be equal to the $service part in the span name.                   | Yes |

Reviewer: 1. In the OpenTracing we had `peer.service` tag, which we used a lot at Uber to validate uninterrupted context propagation. The value for `peer.service` was the actual service name (same one that goes into Resource describing the Tracer), which is often different from `rpc.service`.
2. If the span name is already composed of `{rpc.service}/{rpc.method}`, why are these two attributes _required_?

Reviewer 2: It would also be nice to clarify the relationship between `rpc.service` and `Resource.service`. `Resource.service` would describe the server handling the request, should it match `rpc.service`? If not (e.g., we imagine cases where a single instance handles multiple RPCs to have one resource.service but multiple rpc.service), it'd be a useful example to add either way.

Author: @yurishkuro 
1. Could you elaborate on that please?
2. This is explained in the PR description, second paragraph.

Author: @anuraaga You're referring to server-side spans for incoming calls, right?
From a client perspective, its `service` resource attribute wouldn't have any association with the RPC service it calls within this span.

Reviewer 3: @yurishkuro That is too bad. I introduced `rpc.service` as part of the `net.*` / `peer.*` cleanup in #349 and it was meant as an exactly equivalent replacement for `peer.service`, just specifically for rpc. However, there previously was no proper definition of peer.service and it seems you think the definition here is not equivalent?

Reviewer 2: @arminru Yeah for server spans. Just an example would be nice

Reviewer: Let's take `jaeger-collector` as example. This is the service name that would go into `Resource.service`, it's also the service name as known to my deployment system. The `rpc.service` that collector may expose is `Sampling`, with methods like `GetSamplingStrategy` and `RefreshSamplingStrategy`. `rpc.service` == interface.

This may be different interpretation from what you meant, but the description says `rpc.service` is the `$service` in the span name, which in this case would be `(io.jaergertracing.)Sampling::GetSamplingStrategy`, not `jaeger-collector`.

Author: @anuraaga I added a note distinguishing `rpc.service` (span) and `service.name` (resource) as well as an example below the attributes table, PTAL.

| `rpc.method`   | The name of the method being called, must be equal to the $method part in the span name. | Yes |
| `net.peer.ip`   | See [network attributes][]. | See below |
| `net.peer.name` | See [network attributes][]. | See below |
| `net.peer.port` | See [network attributes][]. | See below |
| `net.transport` | See [network attributes][]. | See below |

At least one of [network attributes][] `net.peer.name` or `net.peer.ip` is required.

> For client-side spans `net.peer.port` is required if the connection is IP-based (it describes the server port they are connecting to).

Reviewer: I would make it recommended, not required. Counterexample: https://google.com - no port number. So if the instrumentation were applied at some generic layer like URLConnection, we're making the instrumentation to _guess_ the port number.

Reviewer 3: An generic URLConnection is no RPC service. You should use the HTTP conventions, not the RPC conventions.

However, there assigning 80 to a port for HTTP URLs and 443 for HTTPS is still not guessing, I think this is surely covered by some RFC.

Author: If it is TCP or UDP via IP there will always be a port number. The URL that you pass to URLConnection (if you're referring to Java here) will throw if you don't specify a protocol. If the protocol is specified but no port, the default port for that protocol will be used and this one can simply be queried by the instrumentation, which needs to `URLConnection.getURL()` anyway to fill the other fields. No guesswork required.

Apart from that, this PR doesn't add the requirement but actually relax it (since it only makes sense for TCP/UDP connections).

Reviewer: @Oberon00 @arminru all you're saying is correct, but can be just as well done by the tracing backend. I am not convinced that _requiring_ the port in this case is beneficial, as it only puts extra burden on the instrumentation that runs on the hot path of the application - the path that we must aim to make as lightweight as possible.

Author: @yurishkuro I think there is less guesswork involved when the instrumentation takes the default port as determined by the library being instrumented rather than guessing it on the tracing backend. However, I changed the wording to not require a port when there is none available, given that it will be available in most cases anyway.

<!-- toc -->

- [Common remote procedure call conventions](#common-remote-procedure-call-conventions)
  * [Span name](#span-name)
  * [Attributes](#attributes)
    + [Service name](#service-name)
  * [Distinction from HTTP spans](#distinction-from-http-spans)
- [gRPC](#grpc)
  * [Status](#status)
  * [Events](#events)

<!-- tocstop -->

> ## Common remote procedure call conventions

Reviewer 4: `Common RPC conventions`?

Author: Do you suggest to shorten the heading or are you questioning the wording?

A remote procedure calls is described by two separate spans, one on the client-side and one on the server-side.

For outgoing requests, the `SpanKind` MUST be set to `CLIENT` and for incoming requests to `SERVER`.

Remote procedure calls can only be represented with these semantic conventions, when the names of the called service and method are known and available.

### Span name

The _span name_ MUST be the full RPC method name formatted as:

```
$package.$service/$method
```

(where $service must not contain dots)

> If no package applies or it is known, the `$package.` part (including the period) is omitted.

Reviewer: ```suggestion
If package name does not apply or is unknown, the `$package.` part (including the period) is omitted.
```

Examples of span names:

- `grpc.test.EchoService/Echo`
- `com.example.ExampleRmiService/exampleMethod`
- `MyCalcService.Calculator/Add` reported by the server and  
  `MyServiceReference.ICalculator/Add` reported by the client for .NET WCF calls
- `MyServiceWithNoPackage/theMethod`

### Attributes

| Attribute name |                          Notes and examples                            | Required? |
| -------------- | ---------------------------------------------------------------------- | --------- |
| `rpc.system`   | A string identifying the remoting system, e.g., `"grpc"`, `"java_rmi"` or `"wcf"`.       | Yes |
| `rpc.service`  | The full name of the service being called, including its package name, if applicable.    | Yes |

> | `rpc.method`   | The name of the method being called, must be equal to the $method part in the span name. | Yes |

Reviewer 4: Do we really need these conventions? Looks like duplicate information with the name. I would definitely not make them required.

Author: `rpc.system` is not part of the span name
`rpc.service` was already there, I think it was taken over from OpenTracing where it was called `peer.service`
`rpc.method` is new

While the latter two are part of the suggested span name, I'd rather have them available without having to parse the name and decompose the components.
On all other semantic conventions the most valuable data, which is used to detect and distinguish services, is provided using attributes and not from parsing the span name, even though this results in some duplication.
I deem the span name less reliable since it's intended as a grouping key and/or to provide a human-readable friendly name rather than a definite, structured data format. I'd expect deviations from the span name guidelines defined in the spec more frequently than not adhering to the attributes defined therein. For example, one could add a "RPC call to xy" prefix to the span name above to "make it look nicer in the UI" even though this would not be allowed as per the spec.

Reviewer 4: Not sure I can buy the "send duplicate data" to avoid "parsing". How do we know which one is more expensive? I would simply not make them required.

Author: Avoiding parsing is one of three reasons I stated in my comment above.
The main reason is that it's in my opinion the least reliable data point we have on a span. All other semantic conventions say something along "_you should use this span name or alternatively that but you can also use something else if it suits better or is of lower cardinality_", so I personally doubt that users/instrumenters will adhere 100% to the guidelines laid out regarding span names. As long as it still makes a good, low-cardinality grouping key, this would be mostly fine, however.
For detecting services from span, I certainly prefer well-defined unambiguous attributes over something that is the main identifier showing up in the UI. The human-readable span name in the UI might, IMHO, be more likely to be populated with something else, because it's more favorable, self-explaining, human-readable or otherwise preferred by users looking at the UI (even if the spec would be pretty strict about it in this particular semantic convention).

Reviewer 4: My fight is still that this should not be required. I am fine and would approve this change if that is removed.

Reviewer 4: I feel that `required` means will not be able to consume data without this field, which is not the case for this.

Author: Alright, I'll make it recommended instead of required and we can revisit it if #653 brings any applicable changes to the definition of required.

Author: @bogdandrutu I made `rpc.method` not required. Should I also remove the requirement for `rpc.service` that is already in place in the current spec version or leave it as it is?

Reviewer 4: Yes please `rpc.service` should be the same.

OffTopic: Probably we should have 3 options (we can discuss in an issue `OPTIONAL`, `RECOMMENDED`, `REQUIRED`) then we should have very few `REQUIRED.

Author: @bogdandrutu done.

Your proposal sounds reasonable, we should discuss this on #653.

#### Service name

On the server process receiving and handling the remote procedure call, the service name provided in `rpc.service` does not necessarily have to match the [`service.name` resource attribute][]. One process can expose multiple RPC endpoints and thus have multiple RPC service names. From a deployment perspective, as expressed by the `service.*` resource attributes, it will be treated as one deployed service with one `service.name`.

As an example, given a process deployed as `QuoteService`, this would be the name that goes into the `service.name` resource attribute which applies to the entire process.
This process could expose two RPC endpoints, one called `CurrencyQuotes` (= `rpc.service`) with a method called `getMeanRate` (= `rpc.method`) and the other endpoint called `StockQuotes`  (= `rpc.service`) with two methods `getCurrentBid` and `getLastClose` (= `rpc.method`).

[network attributes]: span-general.md#general-network-connection-attributes
[net.transport]: span-general.md#nettransport-attribute
[`service.name` resource attribute]: ../../resource/semantic_conventions/README.md#service

### Distinction from HTTP spans

> HTTP calls can generally be represented using [HTTP spans](./http.md) only.

Reviewer 3: ```suggestion
HTTP calls can generally be represented using just [HTTP spans](./http.md).
```

To make one less likely to think of this as "one must not use anything other than HTTP conventions for HTTP calls"

> If they address a particular remote service and method known to the caller, i.e., when it is a remote procedure call transported over HTTP, the `rpc.*` attributes might be added additionally on that span, or in a separate RPC span surrounding the transporting HTTP call.

Reviewer 3: ```suggestion
If they address a particular remote service and method known to the caller, i.e., when it is a remote procedure call transported over HTTP, the `rpc.*` attributes might be added additionally on that span, or in a separate RPC span that is a parent of the transporting HTTP call.
```

(where $service must not contain dots)

> If the package name does not apply or is unknown, the `$package.` part (including the period) is omitted.

Reviewer 3: What does it mean for the package name to apply? To what? Probably you mean:

```suggestion
If there is no package name or if it is unknown, the `$package.` part (including the period) is omitted.
```

> If the package name does not apply or is unknown, the `$package.` part (including the period) is omitted.

Reviewer 3: Actually, I would require a leading period in that case for disambiguation.

Author: > "$service must not contain dots"

This makes it unambiguous already. The leading period would also make it look a bit odd in the UI IMHO.

Reviewer 3: If we clarify that the method name must also not contain dots, then I'm fine with not using a leading dot.

Reviewer 3: Actually, no we don't need that. We use a slash later. Then we should instead clarify that the service name must contain neither dots nor slashes.

Reviewer 3: I see you instead required method to not include slashes, which seems actually more sensible 👍
